spring-cloud-gcp-starters/spring-cloud-gcp-starter-logging/src/main/java/org/springframework/cloud/gcp/logging/autoconfig/StackdriverLoggingAutoConfiguration.java

```diff
@@ -16,28 +16,71 @@
 package org.springframework.cloud.gcp.logging.autoconfig;
 
 import com.google.cloud.logging.TraceLoggingEnhancer;
+import com.google.common.collect.ImmutableList;
 
 import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.cloud.gcp.logging.CompositeTraceIdExtractor;
 import org.springframework.cloud.gcp.logging.LoggingWebMvcConfigurer;
+import org.springframework.cloud.gcp.logging.TraceIdExtractor;
 import org.springframework.cloud.gcp.logging.TraceIdLoggingWebMvcInterceptor;
+import org.springframework.cloud.gcp.logging.XCloudTraceIdExtractor;
+import org.springframework.cloud.gcp.logging.ZipkinTraceIdExtractor;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 import org.springframework.context.annotation.Import;
 
 /**
  * @author Mike Eltsufin
+ * @author Chengyuan Zhao
  */
 @Configuration
 @ConditionalOnClass({ TraceIdLoggingWebMvcInterceptor.class, TraceLoggingEnhancer.class })
+@EnableConfigurationProperties({ StackdriverLoggingProperties.class })
 @ConditionalOnProperty(value = "spring.cloud.gcp.logging.enabled", matchIfMissing = true)
 @Import(LoggingWebMvcConfigurer.class)
 public class StackdriverLoggingAutoConfiguration {
 
 	@Bean
 	@ConditionalOnMissingBean
-	public TraceIdLoggingWebMvcInterceptor getLoggingWebMvcInterceptor() {
-		return new TraceIdLoggingWebMvcInterceptor();
+	public TraceIdLoggingWebMvcInterceptor getLoggingWebMvcInterceptor(
+			TraceIdExtractor extractor) {
+		return new TraceIdLoggingWebMvcInterceptor(extractor);
+	}
+
+	@Bean
+	@ConditionalOnMissingBean
+	public TraceIdExtractor getTraceIdExtractor(
+			StackdriverLoggingProperties loggingProperties) {
+		TraceIdExtractorType type = loggingProperties.getExtractorType();
+		TraceIdExtractorType checkedCombination = type == null
+				? TraceIdExtractorType.XCLOUD_ZIPKIN
+				: type;
+		CompositeTraceIdExtractor extractor;
+		switch (checkedCombination) {
+		case XCLOUD:
```

Comment:
I might be missing something, but it seems to me this is more complex than required.
Just feed a set of headers to `TraceIdLoggingWebMvcInterceptor`?

Reply:
Well, as I pointed above, on the JavaDocs comment we might end up with the requirements to extract trace id from some other place, not just HTTP headers.

I might be to abstract, of course, but that feels like the purpose, @joaoandremartins .
Also pay attention to the difference in the implementations of the `TraceIdExtractor`. We can't do that custom logic having just headers configured.

Reply:
If that's the end goal, sure.
Just for self reference, what other sources could we check?